nifpga/bitfile.py

```diff
@@ -75,27 +84,480 @@ def base_address_on_device(self):
+    type = type_xml.find("SubType")
+    if type is not None:
+        type_name = type.text
+        name = ""
+    else:
+        type_name = type_xml.tag
+        name = type_xml.find("Name").text
+        if name is None:
+            name = ""
+
+    if type_name == "Boolean":
+        return _Bool(name)
+    if type_name == "Cluster":
+        return _Cluster(name, type_xml)
+    if type_name == "FXP":
+        return _FXP(name, type_xml)
+    if type_name == "Array":
+        return _Array(name, type_xml)
+    if type_name == "SGL" or type_name == "DBL":
+        return _Float(name, type_name)
+    if type_name == "CFXP":
+        raise UnsupportedTypeError("The FPGA Interface Python API does not yet support Complex Fixed Point")
+    else:
+        return _Numeric(name, type_name)
+
+
+class _Numeric(object):
+    """ Handles packing and unpacking Numerics such as U8, I8, EnumU8, etc"""
+    def __init__(self, name, type_name):
+        self._name = name
+        if "U8" in type_name:
+            self._size_in_bits = 8
+            self._signed = False
+            self._datatype = DataType.U8
+        elif "I8" in type_name:
+            self._size_in_bits = 8
+            self._signed = True
+            self._datatype = DataType.I8
+        elif "U16" in type_name:
+            self._size_in_bits = 16
+            self._signed = False
+            self._datatype = DataType.U16
+        elif "I16" in type_name:
+            self._size_in_bits = 16
+            self._signed = True
+            self._datatype = DataType.I16
+        elif "U32" in type_name:
+            self._size_in_bits = 32
+            self._signed = False
+            self._datatype = DataType.U32
+        elif "I32" in type_name:
+            self._size_in_bits = 32
+            self._signed = True
+            self._datatype = DataType.I32
+        elif "U64" in type_name:
+            self._size_in_bits = 64
+            self._signed = False
+            self._datatype = DataType.U64
+        elif "I64" in type_name:
+            self._size_in_bits = 64
+            self._signed = True
+            self._datatype = DataType.I64
         else:
-            return Register(xml)
+            raise UnsupportedTypeError("Unrecognized type encountered: %s.  Consider opening an issue on github.com/ni/nifpga" % type_name)
 
-    def _is_register_fxp(self, reg_xml):
-        datatype = reg_xml.find("Datatype")
-        if datatype.find("Array") is not None:
-            datatype = datatype.find("Array").find("Type")
-        for child in datatype.getchildren():
-            if str(DataType.Fxp).lower() not in child.tag.lower():
-                return False
+        self._data_mask = (1 << self._size_in_bits) - 1
+
+        def unpack_numeric_unsigned(bits_from_fpga):
+            data = bits_from_fpga & self._data_mask
+            return data
+
+        signed_bit_mask = 1 << (self._size_in_bits - 1)
```

Comment:
This seems to be only used inside unpack_numeric_signed, so why not move this into that function?

Reply:
Defining it before the function means we calculate the result and then the function just uses that result every time its called.

If I moved it into the function, it would re-calculate it on every call to the function.